add in groups

gowerc opened the pull request:
In this PR I added:
- Support for multi folder workspaces
- Added categorisation for "Build" and "Test" commands
- Added dynamic processing to only load tasks if in a project containing a `DESCRIPTION` file
- Changed the Task definition so it can be written purely in R code 

For the last 2 point please see my comment below

EDIT: Just to say I'm not precious about the formatting or engineering so feel free to change completely if you don't like the way I've styled / laid it out!

gowerc committed:
add in groups

gowerc committed:
added conditional tasks

gowerc commented:
Ok, just added another commit which I think should cover conditioning the tasks on being inside of an R project (based on the existence of a DESCRIPTION file).

gowerc committed:
added multi-workspace support; allowed r code execution

gowerc committed:
fixed minor bug

gowerc committed:
fixed 'install' being classed as a test

gowerc commented:
@renkun-ken ,

Sorry one final change from me here.  I've made a few significant changes. 

- Enabled multi workspace support. Tasks are now evaluated per each available workspace and are added depending on if that workspace contains a `DESCRIPTION` file.  The screenshot below shows me running the `Run Test Task` command in a workspace with 3 folders. For reference `vscoreRTest2` does not contain a `DESCRIPTION` file but the other two do resulting in a `Test` and `Check` command being generated only for `vscoreRTest` and `vscoreRTest3`

![image](https://user-images.githubusercontent.com/8447209/154068345-2019613f-6f27-4964-b864-277b618c351d.png)


- I changed the task that we are defining to embed the `Rscript`. At the moment there as nothing special/unique about our R task, it was just a generic general purpose command line execution.  My understanding of exposing tasks is that it allows you to insert domain specific processing so here I essentially made our tasks wrappers around `Rscript`. This means as a user I can now define my own R specific tasks purely in terms of the R code I want to run i.e.:

![image](https://user-images.githubusercontent.com/8447209/154067769-8a3723f1-0b24-47f4-b696-5f8a3886a68a.png)

renkun-ken commented:
Thanks for working on this! The changes make very good sense.

>Enabled multi workspace support.

The multi-root workspace space support is great!

>At the moment there as nothing special/unique about our R task, it was just a generic general purpose command line execution. My understanding of exposing tasks is that it allows you to insert domain specific processing so here I essentially made our tasks wrappers around Rscript. This means as a user I can now define my own R specific tasks purely in terms of the R code I want to run

I was considering this too. The `command` and `args` task definition is nothing more than a general shell/process task. However, the following might be demanded:

* `Rscript -f file arg1 arg2 ...`
* `Rscript -e expr arg1 arg2`
* `R CMD ...`
* `/path/to/other/Rscript -e expr` (as requested at https://github.com/REditorSupport/vscode-R/issues/948)

I'm not sure if there is a natural way to allow these usages.

renkun-ken commented on src/tasks.ts:
```diff
@@ -1,87 +1,140 @@
+        TYPE,
+        new vscode.ProcessExecution(
+            'Rscript',
+            ['-e', meta.definition.code.join('; ')],
```
Is there a difference between supplying `Rscript -e "expr1; expr2"` and supplying multiple `-e expr` such as `Rscript -e "expr1" -e "expr2"`.

gowerc replied:
I don't think there's any difference but switched it to multiple `-e`'s as I imagine that's probably safer than injecting line breaks.

renkun-ken commented on src/tasks.ts:
```diff
@@ -1,87 +1,140 @@
+        
+        for (const folder of folders) {
+            const is_r_workspace = fs.existsSync(path.join(folder.uri.fsPath, 'DESCRIPTION'));
+            if (is_r_workspace) {
```
`DESCRIPTION` is used to identify workspaces that contain an R package. Should it be named `isRPackage`?

gowerc replied:
Changed

gowerc committed:
renamed variable

gowerc committed:
convert to use -e

gowerc commented:
EDIT: Removed original message:

I was hitting a problem in that you are not allowed to change the definition of a user defined task. I was trying to insert default arguments if the user had left a field blank but apparently it just silently errors if you do that -.-

gowerc commented:
> I was considering this too. The command and args task definition is nothing more than a general shell/process task. However, the following might be demanded: .....

My current thinking is that if the user wants tasks to run arbitrary commands then they can just write those tasks themselves. I don't see the value getting the extension to be a proxy for that if its not actually going to fill in any of the missing components. I quite liked the idea of having our default task be a way for users to specify R code as that's actually adding a service. I'm also thinking that if we need to support more things like `R CMD` we could add a second task type to be exported ? 

> /path/to/other/Rscript -e expr  #948 

I wasn't really sure how best to handle this, I'm kindof against having a popup box ask for the R version as I'm willing to bet that >95% of R users only have 1 version of R installed so it would be a hassle for them. I was wondering instead if it would make more sense to go the same route as python where you can click the bar on the bottom left which then lauches a popup to select the desired interpreter:
![image](https://user-images.githubusercontent.com/8447209/154168440-04254540-a803-4afe-97e6-da8d9d22dbfd.png)

gowerc commented:
Ok, apparently my problem was that you are not allowed to change the definition of a user defined task. I was trying to insert default arguments if the user had left a field blank but apparently it just silently errors if you do that -.-  Will push a new commit shortly

gowerc committed:
added dynamic terminal + terminal args

gowerc commented:
Ok sorry last message and commit (pending review feedback) I have some additional changes that I want to make to the pattern matchers but I will wait until this has been merged before I work on that as I'm concious this is already becoming quite a big change.

In the above commit 316b6b2  I added the following:
- Got tasks to use the R interpreter specified by the user settings
- Added additional `args` argument which allows the user to put additional options in the call to the `R` interpreter (defaults are set to mirror what is used by `Rscript`)
- Added default null `[]` patternMatcher to most tasks in order to suppress it from asking the user to select "do not scan for errors" each time (plan to add some actual pattern matching later on).
- Tested everything to make sure it plays well in multi workspace environments


This image below shows a user defined task taking advantage of our `R` task definition specifying arbitrary R code. Shows that the `arg` options are also correctly passed on. The command fails because I purposefully set my terminal R path to a junk value to show that it will use what the user has specified. It also shows that user defined tasks aren't ignore if the folder doesn't contain a `DESCRIPTION` file
![image](https://user-images.githubusercontent.com/8447209/154237132-f2a4fa73-7d3d-4ac4-87d5-046449c29507.png)


This image just shows I haven't broken any of the prior changes, i.e. this shows that when executing "run build tasks" that it correctly identifies workspace 1 & 3 as being R packages and ignores workspace 2. 
![image](https://user-images.githubusercontent.com/8447209/154237567-cf284b81-cbb9-43f7-8330-83fad5e170d1.png)

renkun-ken commented on src/tasks.ts:
```diff
@@ -3,92 +3,120 @@
-    ) {}
-}
+const getRterm = async function ():Promise<string> {
+    const termOptions = await makeTerminalOptions();
```
We should not use `r.rterm.*` here, it is used for interactive terminal (e.g. radian). We should only use `r.rpath.*` in tasks.

renkun-ken commented on src/tasks.ts:
```diff
@@ -122,24 +147,21 @@ export class RTaskProvider implements vscode.TaskProvider {
-                for (const rtask of rtasks) {
-                    const task = asTask(folder, rtask);
+                for (const rtask of RTasksMeta) {
+                    const task = await asRTask(folder, rtask);
```
Does it make more sense to only `getRPath` once before the loop first? I don't see a need to await each task to get the same rpath.

gowerc replied:
@renkun-ken , You are right. If you do make the changes just be sure to also call `getRPath` in the resolve function as well as the provide function. Perhaps add it as an additional argument to `asRTask`

renkun-ken commented:
I'll directly make some changes to address these problems.

renkun-ken committed:
Refine code

renkun-ken commented on package.json:
```diff
@@ -1674,20 +1674,29 @@
+            "description": "R code to be executed"
           },
-          "args": {
+          "options": {
```
There are multiple terms that might look confusing: args, code args, etc.

Let's use their official terms:

```
$ R --help
Usage: R [options] [< infile] [> outfile]
```

gowerc commented on src/tasks.ts:
```diff
@@ -90,46 +83,35 @@ const RTasksMeta: RTaskMeta[] = [
-    const RtermArgs = zipRTermArguments(meta.definition.code, meta.definition.args ?? []);
-
+function asRTask(rPath: string, folder: vscode.WorkspaceFolder | vscode.TaskScope, info: RTaskInfo): vscode.Task {
+    const args = makeRArgs(info.definition.options ?? defaultOptions, info.definition.code);
```
```
info.definition.options ?? defaultOptions
```

Urgh, I'm angry at myself for not thinking of this. I kept trying to insert the defaultOptions into  `info.definition` which would cause user-defined tasks to fail to resolve. Just inserting it here in the call to `makeRArgs` solves that problem -.-

gowerc commented:
LGTM
